zebra-chain/src/transaction/sighash.rs

```diff
@@ -0,0 +1,289 @@
+#![allow(dead_code, unused_variables)]
+use super::{Transaction, TransparentInput};
+use crate::{
+    parameters::ConsensusBranchId, serialization::ZcashSerialize, types::BlockHeight, Network,
+    NetworkUpgrade,
+};
+use blake2b_simd::Hash;
+use byteorder::{LittleEndian, WriteBytesExt};
+use std::io;
+
+const ZCASH_SIGHASH_PERSONALIZATION_PREFIX: &[u8; 12] = b"ZcashSigHash";
+const ZCASH_PREVOUTS_HASH_PERSONALIZATION: &[u8; 16] = b"ZcashPrevoutHash";
+const ZCASH_SEQUENCE_HASH_PERSONALIZATION: &[u8; 16] = b"ZcashSequencHash";
+const ZCASH_OUTPUTS_HASH_PERSONALIZATION: &[u8; 16] = b"ZcashOutputsHash";
+const ZCASH_JOINSPLITS_HASH_PERSONALIZATION: &[u8; 16] = b"ZcashJSplitsHash";
+const ZCASH_SHIELDED_SPENDS_HASH_PERSONALIZATION: &[u8; 16] = b"ZcashSSpendsHash";
+const ZCASH_SHIELDED_OUTPUTS_HASH_PERSONALIZATION: &[u8; 16] = b"ZcashSOutputHash";
+
+pub const SIGHASH_ALL: u32 = 1;
+const SIGHASH_NONE: u32 = 2;
+const SIGHASH_SINGLE: u32 = 3;
+const SIGHASH_MASK: u32 = 0x1f;
+const SIGHASH_ANYONECANPAY: u32 = 0x80;
+
+pub struct SigHasher<'a> {
+    pub(crate) trans: &'a Transaction,
+    pub(crate) hash_type: u32,
+    pub(crate) network: Network,
+    pub(crate) height: BlockHeight,
+}
+
+impl<'a> SigHasher<'a> {
+    pub fn sighash(self) -> Hash {
+        use NetworkUpgrade::*;
+        let mut hash = blake2b_simd::Params::new()
+            .hash_length(32)
+            .personal(&self.personal())
+            .to_state();
+
+        match self.network_upgrade() {
+            Genesis => unimplemented!(),
+            BeforeOverwinter => unimplemented!(),
+            Overwinter | Sapling => self
+                .hash_sighash_zip143(&mut hash)
+                .expect("serialization into hasher never fails"),
+            Blossom => unimplemented!(),
+            Heartwood => unimplemented!(),
+            Canopy => unimplemented!(),
```

Comment:
Suggestion: Unless there are consensus rule updates, post-Sapling network upgrades will use exactly the same code as Sapling.

Reply:
there are 3 zips that modify the logic of the sighash algo

Reply:
> there are 3 zips that modify the logic of the sighash algo

Those ZIPs are:

Overwinter:
* ZIP 143: Transaction Signature Validation for Overwinter https://github.com/zcash/zips/blob/master/zip-0143.rst
* ZIP 202: Version 3 Transaction Format for Overwinter https://github.com/zcash/zips/blob/master/zip-0202.rst

Sapling:
* ZIP 243: Transaction Signature Validation for Sapling https://github.com/zcash/zips/blob/master/zip-0243.rst

Blossom, Heartwood, and Canopy transaction signatures are exactly the same as Sapling, as documented in their network upgrade ZIPs:
* https://github.com/zcash/zips/blob/master/zip-0206.rst#backward-compatibility
* https://github.com/zcash/zips/blob/master/zip-0250.rst#backward-compatibility
* https://github.com/zcash/zips/blob/master/zip-0251.rst#backward-compatibility